sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala

```diff
@@ -379,11 +381,12 @@ trait FileSourceScanLike extends DataSourceScanExec {
   @transient
   protected lazy val pushedDownFilters = {
     val supportNestedPredicatePushdown = DataSourceUtils.supportNestedPredicatePushdown(relation)
-    // `dataFilters` should not include any metadata col filters
+    // `dataFilters` should not include any constant metadata col filters
     // because the metadata struct has been flatted in FileSourceStrategy
-    // and thus metadata col filters are invalid to be pushed down
+    // and thus metadata col filters are invalid to be pushed down. Metadata that is generated
+    // during the scan can be used for filters.
     dataFilters.filterNot(_.references.exists {
-      case FileSourceMetadataAttribute(_) => true
+      case FileSourceConstantMetadataAttribute(_) => true
```

Comment:
do we have tests for filters with generated metadata cols?

Reply:
row_index uses FileSourceGeneratedMetadataAttribute. Let me double check that a test hits this code path

Reply:
I actually did end up finding a problem here: The filters reference the attributes before they are transformed / flattened, so they are all of type FileSourceMetadataAttribute. I will have to see how I can fix that